File: Demo.Client/Components/Chat/MessageThread.razor.js
import { createElement, createRoot, useState, MessageThread } from '/communication-react.js';

// React component which contains a MessageThread component.
const messageThreadComponent = (props) => {
    const [messages, setMessages] = useState([]);

    props.messages = messages;
    props.onLoadPreviousChatMessages = (messagesToLoad) => onLoadPreviousChatMessages(messagesToLoad, setMessages, props.eventCallback);

    return createElement(MessageThread, props, null);
};

// JS function called to retrieve previous chat messages.
async function onLoadPreviousChatMessages(messagesToLoad, setMessages, eventCallback) {

    // Call the .NET method "OnLoadPreviousChatMessagesAsync()" method.
    var result = await eventCallback.invokeMethodAsync('OnLoadPreviousChatMessagesAsync', messagesToLoad);

    result.messages.map((message) => {
        // Fix the 'createdOn' datetime format from ISO 8601 to JS Date.
        message.createdOn = new Date(message.createdOn);
    });

    setMessages((previousMessages) => {
        return [...result.messages, ...previousMessages];
    });

    return result.noMoreMessages;
};

// Called by Blazor component to render the component using React.
export function initialize(divElement, args, eventCallback) {
    createRoot(divElement).render(createElement(messageThreadComponent, { ...args, eventCallback }, null));
}

File: Demo.Client/Components/Chat/MessageThread.razor.cs
//-----------------------------------------------------------------------
// <copyright file="MessageThread.razor.cs" company="P.O.S Informatique">
//     Copyright (c) P.O.S Informatique. All rights reserved.
// </copyright>
//-----------------------------------------------------------------------

namespace PosInformatique.Azure.Communication.UI.Blazor
{
    using System.Text.Json.Serialization;
    using System.Threading.Tasks;
    using Microsoft.AspNetCore.Components;
    using Microsoft.JSInterop;

    public partial class MessageThread
    {
        private ElementReference messageThreadContainer;

        private JsInterop jsInterop;

        public MessageThread()
        {
            this.jsInterop = new JsInterop(this);

            this.NumberOfChatMessagesToReload = 5;
        }

        [Parameter]
        public string? UserId { get; set; }

        [Parameter]
        public EventCallback<LoadPreviousChatMessagesEvent> OnLoadPreviousChatMessages { get; set; }

        [Parameter]
        public int NumberOfChatMessagesToReload { get; set; }

        [Inject]
        private IJSRuntime JSRuntime { get; set; } = default!;

        protected override async Task OnAfterRenderAsync(bool firstRender)
        {
            if (firstRender)
            {
                await this.jsInterop.InitializeAsync();
            }
        }

        private sealed class JsInterop
        {
            private readonly MessageThread owner;

            private IJSObjectReference? module;

            private DotNetObjectReference<JsInterop>? callbackReference;

            private int startIndex;

            public JsInterop(MessageThread owner)
            {
                this.owner = owner;
                this.callbackReference = DotNetObjectReference.Create(this);
            }

            [JSInvokable]
            public async Task<LoadPreviousChatMessagesResult> OnLoadPreviousChatMessagesAsync(int messagesToLoad)
            {
                return await this.LoadPreviousChatMessagesAsync(messagesToLoad);
            }

            public async Task InitializeAsync()
            {
                if (this.owner.UserId is null)
                {
                    throw new InvalidOperationException($"No '{nameof(this.owner.UserId)}' has been specified on the '{nameof(MessageThread)}' component.");
                }

                this.module = await this.owner.JSRuntime.InvokeAsync<IJSObjectReference>(
                    "import",
                    "./Components/Chat/MessageThread.razor.js");

                var props = new
                {
                    userId = this.owner.UserId,
                    numberOfChatMessagesToReload = this.owner.NumberOfChatMessagesToReload,
                };

                await this.module.InvokeVoidAsync("initialize", this.owner.messageThreadContainer, props, this.callbackReference);
            }

            private async Task<LoadPreviousChatMessagesResult> LoadPreviousChatMessagesAsync(int messagesToLoad)
            {
                var @event = new LoadPreviousChatMessagesEvent(this.startIndex, this.startIndex + messagesToLoad);

                await this.owner.OnLoadPreviousChatMessages.InvokeAsync(@event);

                this.startIndex += messagesToLoad;

                return new LoadPreviousChatMessagesResult(@event.Messages, @event.HasMoreMessages);
            }

            internal sealed class LoadPreviousChatMessagesResult
            {
                public LoadPreviousChatMessagesResult(IReadOnlyList<ChatMessage> messages, bool noMoreMessages)
                {
                    this.Messages = messages;
                    this.NoMoreMessages = noMoreMessages;
                }

                [JsonPropertyName("messages")]
                public IReadOnlyList<ChatMessage> Messages { get; }

                [JsonPropertyName("noMoreMessages")]
                public bool NoMoreMessages { get; }
            }
        }
    }
}
